base/util.go

```diff
@@ -48,24 +48,41 @@ const (
 	kMaxDeltaTtlDuration = 60 * 60 * 24 * 30 * time.Second
 )
 
-// basicAuthURLRegexp is used to match the HTTP basic auth component of a URL
-var basicAuthURLRegexp = regexp.MustCompilePOSIX(`:\/\/[^:/]+:[^@/]+@`)
-
-// RedactBasicAuthURL returns the given string, with a redacted HTTP basic auth component.
-func RedactBasicAuthURL(url string) string {
-	return basicAuthURLRegexp.ReplaceAllLiteralString(url, "://****:****@")
+// RedactBasicAuthURLUserAndPassword returns the given string, with a redacted HTTP basic auth component.
+func RedactBasicAuthURLUserAndPassword(urlIn string) string {
+	redactedUrl, err := RedactBasicAuthURL(urlIn, false)
+	if err != nil {
+		Warnf("%v", err)
+		return ""
+	}
+	return redactedUrl
 }
 
-// basicAuthURLPasswordRegexp is used to match the HTTP basic auth credentials component of a URL
-var basicAuthURLPasswordRegexp = regexp.MustCompile(`:\/\/(?P<username>[^:/]+):(?P<password>[^@/]+)@`)
+// RedactBasicAuthURLPassword returns the given string, with a redacted HTTP basic auth password component.
+func RedactBasicAuthURLPassword(urlIn string) string {
+	redactedUrl, err := RedactBasicAuthURL(urlIn, true)
+	if err != nil {
+		Warnf("%v", err)
+		return ""
+	}
+	return redactedUrl
+}
 
-// RedactBasicAuthURL returns the given string, with a redacted HTTP basic auth password component.
-func RedactBasicAuthURLPassword(url string) string {
-	if basicAuthURLPasswordRegexp.MatchString(url) {
-		redacted := fmt.Sprintf("://${%s}:****@", basicAuthURLPasswordRegexp.SubexpNames()[1])
-		return basicAuthURLPasswordRegexp.ReplaceAllString(url, redacted)
+func RedactBasicAuthURL(urlIn string, passwordOnly bool) (string, error) {
+	urlParsed, err := url.Parse(urlIn)
+	if err != nil {
+		return "", fmt.Errorf("unable to redact URL, returning empty string")
```

Comment:
Adding a comment here to tell us why we're not using the `err` would prevent a future well-intentioned clean-up that wraps err without considering the implications.

```suggestion
		// err contains the unredacted URL, can't wrap or log it
		return "", fmt.Errorf("unable to redact URL, returning empty string")
```